src/main/java/org/example/springproject/repository/book/spec/AuthorSpecificationProvider.java

```diff
@@ -0,0 +1,22 @@
+package org.example.springproject.repository.book.spec;
+
+import java.util.Arrays;
+import org.example.springproject.model.Book;
+import org.example.springproject.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
```

Comment:
Just a suggestion.
What if we create BookSpecificationProvider, where declare all column constants for searching? 
Then, create methods for each specification like
public Specification<Book> authorIn(String[] authors) or public Specification<Book> titles(String[] titles) 

So in BookSpecificationBuilderImpl we can implement something like 
private final BookSpecificationProvider provider;


```
@Override
    public Specification<Book> buildSpecification(BookSearchParametersDto bookSearchDto) {
        return Specification.where(
                        CollectionUtils.isEmpty(bookSearchDto.titles()) ? null :
                                provider.titles(bookSearchDto.titles()))
                .and(CollectionUtils.isEmpty(bookSearchDto.authors()) ? null :
                        provider.authorIn(bookSearchDto.authors()));
    }
```

And use this method in search service method. 
    
So the benefit is: you don't need to create each provider for seraching parameter, maybe you can create some separate method to avoid duplication. Also, all specs will be in one place. 
What do you think about such improvement?

Reply:
but in this case we don't need SpecificationProviderManager, BookSpecificationProviderManager. SpecificationProvider methods won't be universal, i.e. the interface itself is no longer needed. It looks to me a bit like a violation of the Open-Close principle - to add new specifications we will not add a new class, but modify an existing one. I don't see a convenient solution yet, but I'll think about it.

Reply:
It's complitely up to you!